source/common/http/http2/codec_impl.h

```diff
@@ -109,13 +109,21 @@ class ConnectionImpl : public virtual Connection, protected Logger::Loggable<Log
   bool wantsToWrite() override { return nghttp2_session_want_write(session_); }
   // Propagate network connection watermark events to each stream on the connection.
   void onUnderlyingConnectionAboveWriteBufferHighWatermark() override {
-    for (auto& stream : active_streams_) {
-      stream->runHighWatermarkCallbacks();
+    if (h2_watermark_improvements_) {
+      send_pending_frames_cb_->cancel();
+    } else {
+      for (auto& stream : active_streams_) {
+        stream->runHighWatermarkCallbacks();
+      }
     }
   }
   void onUnderlyingConnectionBelowWriteBufferLowWatermark() override {
-    for (auto& stream : active_streams_) {
-      stream->runLowWatermarkCallbacks();
+    if (h2_watermark_improvements_) {
+      send_pending_frames_cb_->scheduleCallbackNextIteration();
```

Comment:
lazy ask: do we have unit tests for 2 backed up sessions both getting "time to write" callbacks?  I'd hope that if the first put the buffer over the limit the second wouldn't serialize and it'd be good to have covreage.

Reply:
Sorry, 2 backed up streams getting "time to write" callbacks from the network.  I think that's doable.

Reply:
I think this is already tested in the integration tests like `DataFlowControlSymmetricStreamConfig`, unless I misunderstood your request.